aesd-char-driver/main.c

```diff
@@ -47,12 +54,54 @@ int aesd_release(struct inode *inode, struct file *filp)
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-	ssize_t retval = 0;
+	ssize_t retval = 0; 
+	struct aesd_dev *dev = filp->private_data;
+	size_t bytes_read = 0;
+	size_t entry_offset = 0;
+	size_t remaining_count = 0;
+	int result_copy;
+	struct aesd_buffer_entry *read_entry;
+
 	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-	/**
-	 * TODO: handle read
-	 */
-	return retval;
+
+	if(mutex_lock_interruptible(&dev->lock))
+	{
+		retval = -ERESTARTSYS;
+		return retval; 
+	}
+
+	read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer_offset, *f_pos, &entry_offset);
+	if(read_entry == NULL) 
+	{
+		retval = 0;
+		goto out;
+	}
+
+	remaining_count = read_entry->size - entry_offset;
+	
+	if(remaining_count > count)
+	{
+		bytes_read = count;
+	}
+	else 
+	{
+		bytes_read = remaining_count;
+	}
+
+	result_copy = copy_to_user(buf, read_entry->buffptr + entry_offset, bytes_read);
+
+	if(result_copy)
+	{
+		retval = -EFAULT;
+		goto out;
```

Comment:
you'll have memory leaks in each of these goto out cases since your out handler isn't handling cleanup of buf